rust/src/io/reader.rs

```diff
@@ -60,7 +77,196 @@ pub async fn read_manifest(object_store: &ObjectStore, path: &Path) -> Result<Ma
+    ///
+    async fn read_primitive_array(&self, field: &Field, batch_id: i32) -> Result<ArrayRef> {
+        let column = field.id;
+        let page_info = self.page_table.get(column, batch_id).ok_or(Error::new(
+            ErrorKind::InvalidData,
+            format!(
+                "field: {} batch {} does not exist in page table",
+                field, batch_id
+            ),
+        ))?;
+        use DataType::*;
+
+        macro_rules! create_plain_decoder {
+            ($a:ty) => {
+                Box::new(PlainDecoder::<$a>::new(
+                    &self.object_reader,
+                    page_info.position,
+                    page_info.length,
+                )?)
+            };
+        }
+
+        let decoder: Box<dyn Decoder> = match field.data_type() {
+            Int8 => create_plain_decoder!(Int8Type),
+            Int16 => create_plain_decoder!(Int16Type),
+            Int32 => create_plain_decoder!(Int32Type),
+            Int64 => create_plain_decoder!(Int64Type),
+            UInt8 => create_plain_decoder!(UInt8Type),
+            UInt16 => create_plain_decoder!(UInt16Type),
+            UInt32 => create_plain_decoder!(UInt32Type),
+            UInt64 => create_plain_decoder!(UInt64Type),
+            Float16 => create_plain_decoder!(Float16Type),
+            Float32 => create_plain_decoder!(Float32Type),
+            Float64 => create_plain_decoder!(Float64Type),
+            _ => {
+                return Err(Error::new(
+                    ErrorKind::InvalidData,
+                    format!("Unsupport primitive type: {}", field.data_type()),
+                ))
+            }
+        };
+        decoder.decode().await
+    }
+
+    async fn read_binary_array(&self, field: &Field, batch_id: i32) -> Result<ArrayRef> {
+        let column = field.id;
+        let page_info = self.page_table.get(column, batch_id).ok_or(Error::new(
+            ErrorKind::InvalidData,
+            format!(
+                "field: {} batch {} does not exist in page table",
+                field, batch_id
+            ),
+        ))?;
+
+        use DataType::*;
+        let decoder: Box<dyn Decoder> = match field.data_type() {
+            Utf8 => Box::new(BinaryDecoder::<Utf8Type>::new(
+                &self.object_reader,
+                page_info.position,
+                page_info.length,
+            )),
+            Binary => Box::new(BinaryDecoder::<BinaryType>::new(
+                &self.object_reader,
+                page_info.position,
+                page_info.length,
+            )),
+            _ => {
+                return Err(Error::new(
+                    ErrorKind::InvalidData,
+                    format!("Unsupported binary type: {}", field.data_type()),
+                ))
+            }
+        };
+        decoder.decode().await
+    }
+
+    /// Read an array of the batch.
+    async fn read_array(&self, field: &Field, batch_id: i32) -> Result<ArrayRef> {
```

Comment:
Should this return a Future rather than await? That way if you're reading a bunch of arrays, you can fire off the requests all at once and await all the futures?

Reply:
`async fn` wraps this function entirely as a future, so in `read_batch`, you can call a bunch of them together, and wait there (at `Schema::fields` level).